Index: test/venn-test.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- test/venn-test.js	(date 1382545674000)
+++ test/venn-test.js	(date 1382546077000)
@@ -216,8 +216,7 @@
           vennSet.not([4,5,6])
           vennSet.should.be.empty
 
-          vennSet.create([1,2,3])
-            .and([]).or([])
+          vennSet = venn.create([1,2,3])
             .not([10,11,12])
             .not([])
 
@@ -271,8 +270,11 @@
             .and([4,1,3])
             .or([2])
             .intersection([4,3,2,1])
+            .not([3,2])
+            .or([1,5])
+            .not([4])
 
-          vennSet.should.eql([1,3,4,2])
+          vennSet.should.eql([1,5])
 
         })
 
@@ -281,22 +283,26 @@
           var noKeyList = venn.create([])
             .union([jane20,jane20,bob30,jane40])
             .or([jane20,eric40])
+            .not([jane20])
             .intersection([jane20,jane20,bob30,jane40,eric40])
+            .not([])
 
-          noKeyList.length.should.equal(4)
-          noKeyList[0].should.eql(jane20)
-          noKeyList[3].should.eql(eric40)
 
+          noKeyList.length.should.equal(3)
+          noKeyList[0].should.eql(bob30)
+          noKeyList[2].should.eql(eric40)
+
           var keyList = venn.create([], keyFunction)
             .or([jane20,jane20,bob30,jane40])
             .union([jane20,eric40])
+            .not([])
             .and([jane20,jane20,bob30,jane40,eric40])
             .union([])
+            .not([jane20,jane40])
 
-          keyList.length.should.equal(3)
-          keyList[0].should.eql(jane20)
-          keyList[1].should.eql(bob30)
-          keyList[2].should.eql(eric40)
+          keyList.length.should.equal(2)
+          keyList[0].should.eql(bob30)
+          keyList[1].should.eql(eric40)
         })
       })
     })
Index: venn.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- venn.js	(date 1382545674000)
+++ venn.js	(date 1382546077000)
@@ -49,6 +49,37 @@
     return this
   }
 
+  var _not = function(set) {
+
+    var that = this
+
+    if(!set || set.length == 0 || !this || this.length == 0) {
+      return this
+    } else {
+
+      var copiedVenn = [].concat(this)
+        , visited = {}
+        , key
+
+      this.length = 0
+
+      set.forEach(function(element) {
+        visited[getKey.call(that,element)] = true
+      })
+
+      copiedVenn.forEach(function(element) {
+
+        key = getKey.call(that,element)
+        if( !visited[key] ) {
+          that.push(element)
+          delete visited[key]
+        }
+      })
+    }
+
+    return this
+  }
+
   var getKey = function(value) {
     if (!this.keyFunction) {
       return bruteForceKeyFunction(value)
@@ -78,6 +109,12 @@
 
   Object.defineProperty(venn_prototype, "and", {
     value : _intersection,
+    writable : false,
+    enumerable : false
+  })
+
+  Object.defineProperty(venn_prototype, "not", {
+    value : _not,
     writable : false,
     enumerable : false
   })
\ No newline at end of file
